crates/amaru-ledger/src/summary/governance/backward_compatibility.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright 2025 PRAGMA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use amaru_kernel::{Epoch, Slot, TransactionPointer};
+use std::collections::BTreeSet;
+
+// In Version 9, the number of dormant epochs depends on the moment the DRep registers *within the
+// epoch*.
+//
+// The original Haskell implementation would keep a counter of dormant epoch in-memory, incremented
+// at epoch boundaries for epochs without proposals. Besides, and unlike Amaru, the Haskell
+// implementation stores the actual expiry epoch for each DRep instead of their registration date
+// and last interaction.
+//
+// The expiration date is then updated for each DRep interations, and "periodically" when the
+// dormant counter is greater than 0. "Periodically" is in fact, **every time a transaction with
+// one or more new proposal is encountered**. So we need to figure out the _last dormant epoch
+// count_ that could be relevant to the DRep.
+pub(crate) fn drep_bonus_mandate(
+    governance_action_lifetime: u64,
+    proposals: &BTreeSet<(TransactionPointer, Epoch)>,
+    (registered_at, registered_in): (Slot, Epoch),
+) -> Epoch {
+    let proposal_until_registration = proposals
+        .iter()
+        .filter(|(_, proposed_in)| proposed_in <= &registered_in)
+        .collect();
+
+    match last_dormant_period(
+        governance_action_lifetime,
+        registered_in,
+        proposal_until_registration,
+    ) {
+        LastDormantEpoch::Empty { .. } | LastDormantEpoch::None { .. } => 0,
+
+        // The dormant epoch is ongoing, so always add the bonus to new DReps;
+        LastDormantEpoch::Last {
+            ending_at: None,
+            length,
+        } => length,
+
+        // The dormant epoch has ended, so only add the bonus if the drep registered *before* it
+        // ended.
+        //
+        // Interestingly, the dormant epoch counter is reset only once all certificates have been
+        // processed, as a last step. This means that if a DRep registers in the same transaction
+        // that a new proposal is submitted, it still get the bonus.
+        // |
+        // * TL; DR -> We need `<=` and not `<`.
+        LastDormantEpoch::Last {
+            ending_at: Some(ending_at),
+            length,
+        } => {
+            if registered_at <= ending_at.slot {
```

Comment:
FWIW this is actually a thing :D

```rust
(registered_at <= ending_at.slot).then_some(5).unwrap_or_default()
```

Reply:
![kaamelott-karadoc-1453962314](https://github.com/user-attachments/assets/d2f61178-c7f4-4e57-a9a7-400275c05ea0)